### ISSUE TITLE
Regression: Salt provisioning fails when box is already provisioned

### ISSUE THREAD

**Issue by martinp** [mitchellh/vagrant#6011]
When running `vagrant provision` on a box that has already been provisioned once, it fails with the following error:

```
$ vagrant provision
==> default: Running provisioner: salt...
Copying salt minion config to vm.
Checking if salt-minion is installed
salt-minion found
Checking if salt-call is installed
salt-call found
Salt binaries found. Configuring only.
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

/tmp/bootstrap_salt.sh  -C

Stdout from the command:



Stderr from the command:

stdin: is not a tty
 * ERROR: In order to run the script in configuration only mode you also need to provide the configuration directory.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
IOError: [Errno 32] Broken pipe
```

Looks like the `bootstrap_salt.sh` script requires `-c` to be given when running in configuration-only mode (`-C`).

Minimal Vagrantfile to reproduce:
```
Vagrant.configure("2") do |config|
  config.vm.box = "chef/ubuntu-14.04"
  config.vm.synced_folder ".", "/vagrant"
  config.vm.provision :salt do |salt|
    salt.minion_config = "salt/minion.yml"
    salt.run_highstate = true
  end
end
```

Vagrant version: 1.7.4

**Comment by martinp** [comment]
This seems to be a regression as it works fine with 1.7.2:

```
$ vagrant provision
==> default: Running provisioner: salt...
Copying salt minion config to vm.
Checking if salt-minion is installed
salt-minion found
Checking if salt-call is installed
salt-call found
Using Bootstrap Options:  -F -c /tmp -C
Salt binaries found. Configuring only.
stdin: is not a tty
<snip>
```

Notice the different bootstrap options: `-F -c /tmp -C`

**Comment by Vanav** [comment]
Is this syntax correct?
```
        if configure && !@machine.config.vm.communicator == :winrm
          options = "%s -F -c %s" % [options, config_dir]
        end
```
It seems that `!(...)` is missing or something like this, and "-c" is never added.
https://github.com/mitchellh/vagrant/commit/4b217e21283c6a17ffbd5f36d622b0934c81e6a8#diff-c4526b51e4e1b1f3f6bff789181cce2b

**Comment by martinp** [comment]
If that's the case, why not just change the condition to `!=` instead?

**Comment by howardroark** [comment]
I can confirm that this happens only with `1.7.3` as well (on both Windows and OSX).  If Salt is installed at all things fall over when trying to provision.  @martinp Did you manage to find any work arounds?

This issue makes it really difficult to work with Salt, and it would be awesome if it could be patched quickly somehow.  Virtualbox is pushing `v5` now and it will be hard to coordinate for people to use the old versions of both Virtualbox and Vagrant in the mean time.

Thanks!

**Comment by martinp** [comment]
@howardroark No. I just reverted to 1.7.2.

**Comment by howardroark** [comment]
As an alternative you can always do...

```
config.vm.provision :shell, path: 'https://raw.githubusercontent.com/saltstack/salt-bootstrap/stable/bootstrap-salt.sh', :args => ""
config.vm.provision :file, source: 'salt/minion', destination: '/etc/salt/minion'
config.vm.provision :shell, inline: 'sudo salt-call state.highstate'
```

The Salt bootstrap script runs every time but it is pretty quick when all the deps are in place.

**Comment by teiresias** [comment]
You need to specify a configuration directory like this:

    salt.bootstrap_options = "-c /tmp"

See bootstrap_salt.sh options...

**Comment by darkn3rd** [comment]
I specified specifying bootstrap_options with `-c /tmp`, `-F -c /tmp`, `-F -c /tmp -C` without success:

```ruby
  config.vm.provision :salt do |salt|
    salt.bootstrap_options = "-F -c /tmp -C"
    salt.run_highstate = true
    salt.verbose = true
  end
```

But I still get the following:
```
DEBUG ssh: Re-using SSH connection.
 INFO ssh: Execute: /tmp/bootstrap_salt.sh  -C (sudo=true)
DEBUG ssh: stderr:  
```

**Comment by darkn3rd** [comment]
Alright, due to bug #6029, with **Ubuntu Precise 12.04**, you need to do the following to work around this issue:
```ruby
  config.vm.provision :salt do |salt|
    # ... 
    salt.bootstrap_options = "-F -c /tmp -P"      # Vagrant Issues #6011, #6029
    # ...
  end
```

I haven't tested if `-F` is exactly needed, but using comments regarding Vagrant 1.7.2 behavior, figured it was good to keep it in there.  The `-P` is required for Ubuntu, as bootstrap.sh uses pip to install Tornado.  The `-P` is NOT needed for **Cent OS 6.6** and **Debian Wheezy**.

**Comment by robob4him** [comment]
+1, da67824 clobbered our salt setup
Thanks @darkn3rd, I'll see if that works for us till this pushes through with a version.

**Comment by ibenschop** [comment]
When I use the bootstrap options solution i keep getting an error that says the F flag does not exist.
It's unfortunate as both of our developers have the same version of vagrant on needs this fix and the other does not. Any ideas?

**Comment by howardroark** [comment]
Hey @ibenschop ... If you can switch to a base Vagrant box without Salt pre-installed you should be able to loose the `-F` flag.  You will need `-P` for Ubuntu to enable pip installation.

Here are some options... https://atlas.hashicorp.com/boxes/search?utm_source=vagrantcloud.com&vagrantcloud=1

**Comment by ibenschop** [comment]
Hi thanks Howard. We did exactly what you said. Removed the preinstalled
salt plugin and it worked across. Much appreciated!

**Comment by mitchellh** [comment]
I believe a fix is in with the latest PR merges. Please verify with Vagrant 1.8 and if the issue persist please reopen. Thanks!

**Comment by pkruithof** [comment]
It's not resolved for me when using Vagrant 1.8.1, same error as in #6029
